pyqtorch/utils.py

```diff
@@ -221,6 +222,57 @@ def is_diag(H: Tensor, atol: Tensor = ATOL) -> bool:
     return torch.count_nonzero(torch.abs(offdiag_view).gt(atol)) == 0
 
 
+def finitediff(
+    f: Callable,
+    x: Tensor,
+    derivative_indices: tuple[int, ...],
+    eps: float | None = None,
+) -> Tensor:
+    """
+    Compute the finite difference of a function at a point.
+
+    Args:
+        f: The function to differentiate.
+        x: Input of size `(batch_size, input_size)`.
+        derivative_indices: Which *input* to differentiate (i.e. which variable x[:,i])
+        eps: finite difference spacing (uses `torch.finfo(x.dtype).eps ** (1 / (2 + order))`
+            as default)
+
+    Returns:
+        (Tensor): The finite difference of the function at the point `x`.
+    """
+
+    if eps is None:
+        order = len(derivative_indices)
+        eps = torch.finfo(x.dtype).eps ** (1 / (2 + order))
+
+    # compute derivative direction vector(s)
+    eps = torch.as_tensor(eps, dtype=x.dtype)
+    _eps = 1 / eps  # type: ignore[operator]
+    ev = torch.zeros_like(x)
+    i = derivative_indices[0]
+    ev[:, i] += eps
+
```

Comment:
Since `eps` seems to not be used anymore in the code, maybe having `ev[:, i] := eps` first and then giving a meaningful variable name for `1 / eps` operation (what is even `ev`?) to favor code description

Reply:
Good remarks, I took it from qadence, ev represents the step/delta for differentiation. Will rename accordingly.